internal/text/text.go

```diff
@@ -0,0 +1,284 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package text
+
+import (
+	"bytes"
+	"fmt"
+	"sort"
+	"strconv"
+	"strings"
+	"text/scanner"
+)
+
+const (
+	// FailureFieldFilename references the Filename field of a Failure.
+	FailureFieldFilename FailureField = iota
+	// FailureFieldLine references the Line field of a Failure.
+	FailureFieldLine
+	// FailureFieldColumn references the Column field of a Failure.
+	FailureFieldColumn
+	// FailureFieldID references the ID field of a Failure.
+	FailureFieldID
+	// FailureFieldMessage references the Message field of a Failure.
+	FailureFieldMessage
+)
+
+var (
+	// DefaultFailureFields are the default FailureFields.
+	DefaultFailureFields = []FailureField{
+		FailureFieldFilename,
+		FailureFieldLine,
+		FailureFieldColumn,
+		FailureFieldMessage,
+	}
+
+	_failureFieldToString = map[FailureField]string{
+		FailureFieldFilename: "filename",
+		FailureFieldLine:     "line",
+		FailureFieldColumn:   "column",
+		FailureFieldID:       "id",
+		FailureFieldMessage:  "message",
+	}
+	_stringToFailureField = map[string]FailureField{
+		"filename": FailureFieldFilename,
+		"line":     FailureFieldLine,
+		"column":   FailureFieldColumn,
+		"id":       FailureFieldID,
+		"message":  FailureFieldMessage,
+	}
+)
+
+// FailureField references a field of a Failure.
+type FailureField int
```

Comment:
There looks like a decent amount of boilerplate around mapping these fields from their `iota` representation to their corresponding `string`. Seeing as we only ever really care about the `string` representation, should we consider simply setting these upfront?

```
// FailureField references a field of a Failure.
type FailureField string

const (
  Filename FailureField = "filename"
  Line FailureField = "line"
  Column FailureField = "column"
  ID FailureField = "id"
  Message FailureField = "message"
)
```

Reply:
This feels pretty standard for Golang enums. Using strings allows people to get lazy and cast actual strings instead of the variable constants, which is why ints are used in Golang.

Reply:
Works for me.